src/main/java/com/microsoft/aad/msal4j/AcquireTokenSilentSupplier.java

```diff
@@ -35,11 +37,20 @@ AuthenticationResult execute() throws Exception {
                     silentRequest.parameters().scopes(),
                     clientApplication.clientId());
 
+            if (res == null) {
+                throw new MsalClientException(AuthenticationErrorMessage.NO_TOKEN_IN_CACHE, AuthenticationErrorCode.CACHE_MISS);
+            }
+
             if (!StringHelper.isBlank(res.accessToken())) {
                 clientApplication.getServiceBundle().getServerSideTelemetry().incrementSilentSuccessfulCount();
             }
 
-            if (silentRequest.parameters().forceRefresh() || StringHelper.isBlank(res.accessToken())) {
+            //Determine if the current token needs to be refreshed according to the refresh_in value
+            long currTimeStampSec = new Date().getTime() / 1000;
+            boolean afterRefreshOn = res.refreshOn() != null && res.refreshOn() > 0 &&
+                    res.refreshOn() < currTimeStampSec && res.expiresOn() >= currTimeStampSec;
```

Comment:
res.expiresOn() >= currTimeStampSec means you will not do refresh if AT is expired, why ?

Reply:
I didn't want to change any behavior if the token was expired, so if the token was after the `expiresOn` time then `refreshOn` time is irrelevant and we shouldn't do any of the `refreshOn` behavior. If the token was after the `refreshOn` time, then if any service needed to refresh it was down then the current token should be returned as though it nothing happened, which is not the (existing) behavior for an expired token.